### PR TITLE
Convert facility_ids in staging

### PR DESCRIPTION
**Pull request by Tonksthebear**
## Description of change
Convert `facility_id` to updated 9xx range ids in staging

## Original issue(s)
[department-of-veterans-affairs/va.gov-team#3648](https://app.zenhub.com/workspaces/va-mobile---backend-61d75d03a7505500147ec7f6/issues/department-of-veterans-affairs/va-mobile-app/3648)

## Things to know about this PR
A previous assumption made about this endpoint was a user should never reach it without cached appointments. Thus, there is an error thrown if appointments are blank. This PR updates adds a safety operator to the logic in order to allow allow this custom error to actually be thrown.

The added spec tests for the correct error being thrown. I had to use a regex to get solve the rubocop line length error.

### CODE REVIEW

**Review by RachalCassity — APPROVED** [review]

File: modules/mobile/app/controllers/mobile/v0/facilities_info_controller.rb
Comment (kpethtel) [inline modules/mobile/app/controllers/mobile/v0/facilities_info_controller.rb:45]: TBH, the existing code seems slightly wrong. I'm starting to see what Andrew was talking about with this bug occurring because we're using the API and not the mobile app. This code should work through the mobile app because we pre-cache appointments on login. And we intend this code to be used as part of the appointments flow and it might not be possible to get to this request in the mobile app without first navigating to the appointments index, which will also cache appointments. But we've already added a feature flag to the login pre-cache because we know it might cause performance issues and may need to be turned off. So that's fragile. And if we make a design change that allows the user to navigate to direct schedule without hitting the index first, this becomes broken. 

So yes it would take two points of failure for this to be a problem, but I also think this could be easily fixed by extracting the `fetch_cached_or_service` from the appointments controller into a service object or model method and using it here. So instead of fetching cached here, we just do a cached or fetch. What do you guys think @aherzberg @jperk51 ?

At the very least, I think we should reword this error message to something like, "User does not have cached appointments". But it's also worth pointing out that this might not indicate an error at all. It could just be a user who has no previous appointments, in which case it would probably make sense to just return an empty response.

**Review by kpethtel — COMMENTED** [review]
Your code seems good, but while looking at this I realized there's another issue that I think we should consider. Let's talk it through.

File: modules/mobile/app/controllers/mobile/v0/facilities_info_controller.rb
Comment (aherzberg) [reply modules/mobile/app/controllers/mobile/v0/facilities_info_controller.rb:45]: Jayson and I discussed this a bit when implementing it. I think the idea around just using the cache was that doing an appointment fetch on top of the original request was going to really draw out the response time since appointments is pretty expensive. we were going to use the logs to see how often there wasn't a cache and make an informed decision on how to address it. I personally don't mind putting a fetch in there but It definitely complicates things when you've got appointment fetching happening outside the appointment controller. 

I do agree we can update the wordage of the raised error to say its the cache that's empty.

Comment (kpethtel) [reply modules/mobile/app/controllers/mobile/v0/facilities_info_controller.rb:45]: I'm not sure I understand what you mean about fetching outside the appointments controller complicating things. We already do it in the users controller during the pre-cache. 

I can see the argument that a controlled error is better than a timeout. However, we're treating this as a 404, which no longer seems correct. I think this is more like a 422. Also, I think we should consider the possibility that a user might not have any prior appointments, in which case perhaps an empty response might be better than an error.

Comment (Tonksthebear) [reply modules/mobile/app/controllers/mobile/v0/facilities_info_controller.rb:45]: I too was wondering about the event the user legitimately doesn't have any appointments. I wasn't sure how front end handles our errors, so I suppose I thought they would provide a message to the user if they don't have appointments.

I see this PR was approved, but I'm hesitant to move forward with it without a conclusion being made on if this should be tweaked. Would the proposed changes be a separate ticket, or would it be handled in this existing PR?

File: modules/mobile/spec/request/facilities_info_request_spec.rb
Comment (Tonksthebear) [reply modules/mobile/spec/request/facilities_info_request_spec.rb:175]: Added!

File: modules/mobile/app/controllers/mobile/v0/facilities_info_controller.rb
Comment (kpethtel) [reply modules/mobile/app/controllers/mobile/v0/facilities_info_controller.rb:45]: The person who approved is not a mobile team developer. We just don't have code owner approval set up correctly to block merging until you have approval from a mobile team developer if the code is in the mobile module. 

I think we need to come to a consensus about what we want to do before we merge this. I'll hit you up on slack.

Comment (Tonksthebear) [reply modules/mobile/app/controllers/mobile/v0/facilities_info_controller.rb]: Ok just updated with the logic discussed via slack and added new specs. Ready for review again

**Comment by Tonksthebear** [comment]
Ok just updated with the logic discussed via slack and added new specs. Ready for review again

File: modules/mobile/app/controllers/mobile/v0/facilities_info_controller.rb
Comment (kpethtel) [inline modules/mobile/app/controllers/mobile/v0/facilities_info_controller.rb:41]: I think this needs to be `sort_by!`. I'm also not sure this comment is correct. I believe facilities that are not in the recent appointments (in the case that there are recent appointments) will be at the bottom of the list and sorted by id.

File: modules/mobile/app/controllers/mobile/v0/facilities_info_controller.rb
Comment (kpethtel) [inline modules/mobile/app/controllers/mobile/v0/facilities_info_controller.rb:54]: If we do this in the case of an empty appointments, we just end up sorting facilities by id. Based on our conversation in slack, I was expecting us to sort alphabetically but not log. So more like
```
if appointments.nil?
  # log
  facilities
elsif appointments.empty?
  facilities
else
  ...
end
```

I think the logic here is basically "sort the user's facilities by recency of appointment and if they haven't had an appointment at the facility just tack it on at the end."

File: modules/mobile/spec/request/facilities_info_request_spec.rb
Comment (kpethtel) [inline modules/mobile/spec/request/facilities_info_request_spec.rb:167]: Can we test by facility names instead of ids to demonstrate the alphabetical-ness?

**Review by kpethtel — COMMENTED** [review]
I think there might be a couple of small issues. LMK what you think.

File: modules/mobile/app/controllers/mobile/v0/facilities_info_controller.rb
Comment (jperk51) [reply modules/mobile/app/controllers/mobile/v0/facilities_info_controller.rb:41]: There should be a test that checks to make sure the leftovers are are alpha. It could be that the IDs are resulting in that situation too, but they could easily be edited to ensure it's actually alpha

The comment could be updated to make it clear they are at the bottom of the list which is intended behavior. The list should have recent appointment matches at the top sorted by most recent appointment then any facilities that don't have a match should be below those and sorted alpha.

File: modules/mobile/app/controllers/mobile/v0/facilities_info_controller.rb
Comment (jperk51) [reply modules/mobile/app/controllers/mobile/v0/facilities_info_controller.rb:54]: This is correct. Those tacked on facilities should be sorted alphabetically

File: modules/mobile/spec/request/facilities_info_request_spec.rb
Comment (jperk51) [reply modules/mobile/spec/request/facilities_info_request_spec.rb:167]: Agreed.

Comment (Tonksthebear) [reply modules/mobile/spec/request/facilities_info_request_spec.rb:167]: Sure, I just copied an existing spec for testing alphabetical-ness since those were already "approved". Should I go ahead and change those as well?

File: modules/mobile/app/controllers/mobile/v0/facilities_info_controller.rb
Comment (Tonksthebear) [reply modules/mobile/app/controllers/mobile/v0/facilities_info_controller.rb:41]: I left existing code untouched because my assumption was it was all operating the way it was intended. I'll get this updated

File: modules/mobile/app/controllers/mobile/v0/facilities_info_controller.rb
Comment (Tonksthebear) [reply modules/mobile/app/controllers/mobile/v0/facilities_info_controller.rb:54]: I believe the change to `facilities.sort_by(&:name)` should mean `facilities` is sorted alphabetically right from the beginning. Then code in the `else` block should just return facilities sorted alphabetically because it already returns "leftover" facilities alphabetically. Unless you like the `elsif appointements.empty?` for code clarity

File: modules/mobile/spec/request/facilities_info_request_spec.rb
Comment (kpethtel) [reply modules/mobile/spec/request/facilities_info_request_spec.rb:167]: Based on the conversation here, I'm realizing that the comment probably is wrong and the specs probably are correct. When there are no previous appointments for a facility, it just sorts by id. It makes me wonder if what we really wanted was to sort by name here as a fallback instead of id, like:
`facilities.sort_by { |facility| [rf_hash[facility.id] || recent_facilities.size, facility.name] }`

File: modules/mobile/app/controllers/mobile/v0/facilities_info_controller.rb
Comment (jperk51) [reply modules/mobile/app/controllers/mobile/v0/facilities_info_controller.rb:54]: That was the intended functionality. @Tonksthebear just write or make sure the existing tests are adequately ensuring that's what's happening

**Comment by Tonksthebear** [comment]
Ok this last commit changed quite a bit. I'll comment on specific choices I made to explain

File: modules/mobile/app/controllers/mobile/v0/facilities_info_controller.rb
Comment (Tonksthebear) [inline modules/mobile/app/controllers/mobile/v0/facilities_info_controller.rb:8]: I thought it would be better to define valid sort methods up front so that we can fail earlier in the process. Before, failure would happen after most of the logic was completed. Now, we don't bother doing anything unless the provided sort method is valid

File: modules/mobile/app/controllers/mobile/v0/facilities_info_controller.rb
Comment (Tonksthebear) [inline modules/mobile/app/controllers/mobile/v0/facilities_info_controller.rb:37]: I wanted to move alpha sorting outside of `sort_by_recent_appointments` because I felt it didn't belong in the method due to its naming

File: modules/mobile/app/controllers/mobile/v0/facilities_info_controller.rb
Comment (Tonksthebear) [inline modules/mobile/app/controllers/mobile/v0/facilities_info_controller.rb:65]: This was able to be simplified because we want to leave any facility NOT in `recent_facilities` untouched, as they are already alphabetized from earlier. Now this will just push all non-recent facilities to the bottom

**Review by aherzberg — APPROVED** [review]
LGTM

**Review by kpethtel — CHANGES_REQUESTED** [review]
One blocking issue regarding variable scope.

File: modules/mobile/app/controllers/mobile/v0/facilities_info_controller.rb
Comment (jperk51) [reply modules/mobile/app/controllers/mobile/v0/facilities_info_controller.rb:65]: If the original was sorting by id that was a mistake. The facilities that don't match past appointments should be sorted alphabetically after the ones that do match.

File: modules/mobile/app/controllers/mobile/v0/facilities_info_controller.rb
Comment (jperk51) [reply modules/mobile/app/controllers/mobile/v0/facilities_info_controller.rb:74]: +1 for adding this test

File: modules/mobile/app/controllers/mobile/v0/facilities_info_controller.rb
Comment (Tonksthebear) [reply modules/mobile/app/controllers/mobile/v0/facilities_info_controller.rb:40]: I felt they were different situations. There could be the event where someone adds to `SORT_METHODS` but forgets to actually implement it. So the switch statement reaches its `else` block but doesn't let anyone know it didn't actually do anything. That's separate from the sort method being invalid

File: modules/mobile/app/controllers/mobile/v0/facilities_info_controller.rb
Comment (Tonksthebear) [reply modules/mobile/app/controllers/mobile/v0/facilities_info_controller.rb:74]: Good catch, I'll add a test

File: modules/mobile/app/controllers/mobile/v0/facilities_info_controller.rb
Comment (kpethtel) [reply modules/mobile/app/controllers/mobile/v0/facilities_info_controller.rb:40]: Yeah, that's the "coding defensively" argument I was talking about. Like I said, not requesting a change here. Just talking it through.

**Comment by Tonksthebear** [comment]
@kpethtel I'm not sure why I couldn't respond directly to your [comment](https://github.com/department-of-veterans-affairs/vets-api/pull/10540#discussion_r956139306), but @jperk51 said non-recent facilities it should be sorted alpha as priority in slack. I agree though, confirming here again is important. The original implementation ignored alpha and failed tests

**Comment by Tonksthebear** [comment]
> @kpethtel I'm not sure why I couldn't respond directly to your [comment](https://github.com/department-of-veterans-affairs/vets-api/pull/10540#discussion_r956139306), but @jperk51 said non-recent facilities it should be sorted alpha as priority in slack. I agree though, confirming here again is important. The original implementation ignored alpha and failed tests

Ignore this, weird issues on my end for GitHub that a refresh solved. Didn’t see any of Jayson’s replies

**Review by kpethtel — APPROVED** [review]
It might be nice to separate out all of the sorting-related tests into a separate `describe 'Sorting'` block. 

The problem with these "with feature flag on/off" test blocks is that they push us to duplicate tests. IMO, these tests should be re-written to avoid this duplication. We should do set up a spec for the appointments proxy, which is where the feature flag is used and unit test the relevant methods. 

But that's outside the scope of the work you're doing here. I might spend some time looking at a refactor of some of these specs. Currently, we don't have any unit tests for the appointments proxy and I think it's making the request specs messy. 

But this LGTM. 🚀

**Comment by Tonksthebear** [comment]
> It might be nice to separate out all of the sorting-related tests into a separate `describe 'Sorting'` block.
> 
> The problem with these "with feature flag on/off" test blocks is that they push us to duplicate tests. IMO, these tests should be re-written to avoid this duplication. We should do set up a spec for the appointments proxy, which is where the feature flag is used and unit test the relevant methods.
> 
> But that's outside the scope of the work you're doing here. I might spend some time looking at a refactor of some of these specs. Currently, we don't have any unit tests for the appointments proxy and I think it's making the request specs messy.
> 
> But this LGTM. 🚀

Agreed, it pained me to see all the duplication. If you want me to spend any time on a refactor, let me know!

**Pull request merged**
